### PR TITLE
[TTAHUB-2318] goal & objective collaborators

### PR DESCRIPTION
**Pull request by GarrettEHill**
## Description of change

- Record collaborators for goal & objectives
- Populate all existing goals & objectives
- Collaborator types:
  - Creator - The user who creates a goal or objective
  - Editor - The user(s) who edits the text of a goal or objective
  - Linker - The user who links the goal or objective to a report
  - Utilizer - The user who created report and users listed as collaborators on report.

- added db error log monitoring to circleci builds

## How to test
Create or edit a goal, check the db looking at the GoalCollaborators and ObjectiveCollaborators.

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-2318

### Every PR

- [ ] Meets issue criteria
- [ ] JIRA ticket status updated
- [ ] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions
- [ ] UI review complete

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

File: src/models/helpers/goalCollaborator.js
Comment (nvms) [inline src/models/helpers/goalCollaborator.js:219]: I think this only grabs the first value of the array returned by Object.entries. Is that what you're going for?

Screenshot 2023-11-26 at 12 18 56 PM

File: src/models/helpers/genericCollaborator.js
Comment (nvms) [inline src/models/helpers/genericCollaborator.js:367]: For simplicity sake, can this be refactored? It's pretty tough to follow as-is with the ternaries wrapped in a spread.

Untested simplification:

```javascript
const updatePromises = updates.map(async (update) => {
  await sequelize.models[collaboratorDetails[genericCollaboratorType].collaborators].update(
    { linkBack: update.linkBack },
    {
      where: { id: update.id },
      independentHooks: true,
      ...(transaction && { transaction }),
    },
  );
});

const deletePromise = sequelize.models[collaboratorDetails[genericCollaboratorType].collaborators].destroy({
  where: { id: deletes },
  independentHooks: true,
  ...(transaction && { transaction }),
});

await Promise.all([...updatePromises, deletePromise]);
```

File: src/migrations/20231126165733-goalusers.js
Comment (hardwarehuman) [inline src/migrations/20231126165733-goalusers.js:326]: Shouldn't collaborators default to utilizers rather than linkers?

**Comment by thewatermethod** [comment]
new AR. created a goal + objective in the UI. pressed save draft. 500 response. postgres output:

```
postgres_docker                          | 2023-11-30 18:37:16.083 UTC [43] ERROR:  insert or update on table "GoalCollaborators" violates foreign key constraint "GoalCollaborators_goalId_fkey"
postgres_docker                          | 2023-11-30 18:37:16.083 UTC [43] DETAIL:  Key (goalId)=(111585) is not present in table "Goals".
```

File: .circleci/config.yml
Comment (thewatermethod) [inline .circleci/config.yml:359]: nice

**Comment by thewatermethod** [comment]
When I merge a goal, I become the author of the new goal. I'm not sure that makes sense - as far as the user is concerned, they do not think they are creating a new goal, but picking a goal to be the winner from a set. Wouldn't the expectation be that the author of the "final" goal is the author of the new goal?

**Comment by GarrettEHill** [comment]
> When I merge a goal, I become the author of the new goal. I'm not sure that makes sense - as far as the user is concerned, they do not think they are creating a new goal, but picking a goal to be the winner from a set. Wouldn't the expectation be that the author of the "final" goal is the author of the new goal?

I will look at the requirements again

File: src/migrations/20231126165733-goalusers.js
Comment (hardwarehuman) [inline src/migrations/20231126165733-goalusers.js:365]: Shouldn't this one also filter out cases where the dml_as is in -1 and 0? Otherwise we'll miss cases where pre-audit log ARs get ARGs via migration.

**Comment by thewatermethod** [comment]
Can we document the meanings behind the collaborator types somewhere? (utilizer, linker) They may seem self explanatory but with any of these terms of art, the meaning can break down as new people find them.

File: src/migrations/20231127000000-objectiveusers.js
Comment (hardwarehuman) [inline src/migrations/20231127000000-objectiveusers.js:258]: Like the other comment, I think we want to filter out cases where the dml_as is in -1 and 0 so we don't miss cases where pre-audit log ARs get AROs via migration.

**Review by hardwarehuman — APPROVED** [review]
I think there's probably a change in the migrations but it's very edge-casey so approving.

**Comment by thewatermethod** [comment]
> Can we document the meanings behind the collaborator types somewhere? (utilizer, linker) They may seem self explanatory but with any of these terms of art, the meaning can break down as new people find them.

dang that's right in the PR description.

File: docs/logical_data_model.puml
Comment (hardwarehuman) [inline docs/logical_data_model.puml:257]: I think we should have comments explaining the thinking behind linkbacks

Comment (GarrettEHill) [reply docs/logical_data_model.puml:258]: this is a generated file, so.... no comments here

File: src/goalServices/goals.js
Comment (GarrettEHill) [reply src/goalServices/goals.js:2874]: with comments its should be easier

File: src/models/collaboratorType.js
Comment (GarrettEHill) [reply src/models/collaboratorType.js:39]: its a future proofing construct to allow us to replace definitions without the need to modify existing historical data.

File: src/models/hooks/activityReport.js
Comment (GarrettEHill) [reply src/models/hooks/activityReport.js:1011]: that is not a correct fix, but I will look into the error

File: src/models/hooks/activityReport.js
Comment (GarrettEHill) [reply src/models/hooks/activityReport.js:1019]: not a correct fix

**Comment by thewatermethod** [comment]
I think we might wants tests covering the assigning of a collaborator on a new goal created on AR as well as the merge goal assignments, at least.

File: src/migrations/20231207213618-goalusers.js
Comment (AdamAdHocTeam) [inline src/migrations/20231207213618-goalusers.js:64]: Something we are always lacking in, nice.

File: src/models/helpers/genericCollaborator.js
Comment (AdamAdHocTeam) [inline src/models/helpers/genericCollaborator.js:119]: Assuming I'm reading this correctly is there any concern for 'genericCollaboratorType' coming back undefined or empty then attempting to access a property like 'id'?

File: src/lib/semaphore.js
Comment (AdamAdHocTeam) [inline src/lib/semaphore.js:3]: Are we using the semaphore anywhere currently?

**Review by AdamAdHocTeam — COMMENTED** [review]
Overall this looks good and the migration runs without issue.

Left a few comments.

File: src/models/helpers/genericCollaborator.js
Comment (GarrettEHill) [reply src/models/helpers/genericCollaborator.js:119]: i think it should be good

File: src/lib/semaphore.js
Comment (GarrettEHill) [reply src/lib/semaphore.js:3]: yes there is a race condition that needed to be protected

**Comment by GarrettEHill** [comment]
@thewatermethod nice find, that took the rest of the day. but now that is all fixed.

**Review by thewatermethod — APPROVED** [review]
Everything looks good as far as testing from the UI. Would ask that you please hold off merging until we get the prod PR in later today. 

Also not sure if other reviewers are satisfied, or if this PR was a candidate for a group-discussion, might be worth consulting the team at large

**Comment by thewatermethod** [comment]
I think the encoded LDM might be out of date on this branch? Just noticing that I had uncommitted changes hanging around locally from running this migration earlier

**Comment by thewatermethod** [comment]
> Would ask that you please hold off merging until we get the prod PR in later today.

This has been merged

**Review by AdamAdHocTeam — COMMENTED** [review]
Getting a weird behavior here not sure if it has something to do with impersonating users or my understanding of how its supposed to work.

1. Create goal with two objectives (creator)
2. Update with different user (editor)
3. Assign to a report (linker)
4. Approve with a new user (utilizer)

View the goal collaborator table I see two users as the utilizer. But I don't think they make sense. I would expect to see the approver as the utilizer. Instead I see the creator and one other user that was NOT involved in the process at all?

image

This was the user I approved the report with: 
image


**UPDATED:** Ok this might be a non-issue seems the utilizer is the creator and selected collab on the report. Can you confirm this behavior as the intended.

**Review by AdamAdHocTeam — COMMENTED** [review]
Another question for a possible issue. If for some reason the report is Unlocked do we want to remove the utilized records?

**Comment by GarrettEHill** [comment]
the unapproved report is removed from utilizer, when a report is deleted or a goal removed from a report it is unlinked

**Review by AdamAdHocTeam — APPROVED** [review]

**Pull request merged**
